### ISSUE TITLE
As a developer, I want to investigate, understand, and implement a reader for the WRDS HEFS service

### ISSUE THREAD

**Issue by epag** [NOAA-OWP/wres#138]
---

Author Name: **Hank** (Hank)
Original Redmine Issue: 121930, https://vlab.noaa.gov/redmine/issues/121930
Original Date: 2023-10-26

---

The URL is:

https://nwcal-wrds.[domain]/api/hefs/v1/

Swagger:

https://nwcal-wrds.[domain]/api/hefs/swagger/

I would like to investigate the API and gain an understanding relatively soon. We can then decide if and when we want to write a reader for it. Marking this as high priority, but not urgent,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-10-26T15:58:33Z

---

Fixed a typo.

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-11-02T16:56:56Z

---

We are waiting on this until a user asks for it, but doing an initial investigation sooner makes sense.

This might end up becoming a framework picked up for the other services down the road. 

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-11-02T17:37:41Z

---

Here is an example URL:

https://nwcal-wrds.[domain]/api/hefs/v1/series/?limit=10&start_date=2023-10-01

Here is the example output:

```<code class="json">
{
    "count": 6895131,
    "next": "http://nwcal-wrds.[domain]/api/hefs/v1/series/?limit=10&offset=10&start_date=2023-10-01",
    "previous": null,
    "results": [
        {
            "id": "b939ffe0-a3e4-4dd6-9c0d-d2f325dbf874",
            "created_at": "2023-10-11T18:57:48.449916Z",
            "updated_at": "2023-10-11T18:57:48.449940Z",
            "deleted_at": null,
            "is_deleted": false,
            "type": "instantaneous",
            "location_id": "PDLC2",
            "parameter_id": "QINE",
            "ensemble_id": "MEFP",
            "ensemble_member_index": 1951,
            "time_step_unit": "second",
            "time_step_multiplier": 21600,
            "start_date_date": "2022-05-16",
            "start_date_time": "00:00:00",
            "end_date_date": "2022-06-15",
            "end_date_time": "00:00:00",
            "forecast_date_date": "2022-05-16",
            "forecast_date_time": "00:00:00",
            "miss_val": "nan",
            "station_name": "PDLC2R",
            "lat": 38.4872222222,
            "lon": -105.373055556,
            "x": null,
            "y": null,
            "z": null,
            "units": "CFS",
            "creation_date": "2022-05-16",
            "creation_time": null
        },
        {
            "id": "dc822be7-a136-4845-b472-00f65d5cf7fa",
            "created_at": "2023-10-11T18:57:48.498368Z",
            "updated_at": "2023-10-11T18:57:48.498377Z",
            "deleted_at": null,
            "is_deleted": false,
            "type": "instantaneous",
            "location_id": "PDLC2",
            "parameter_id": "QINE",
            "ensemble_id": "MEFP",
            "ensemble_member_index": 1952,
            "time_step_unit": "second",
            "time_step_multiplier": 21600,
            "start_date_date": "2022-05-16",
            "start_date_time": "00:00:00",
            "end_date_date": "2022-06-15",
            "end_date_time": "00:00:00",
            "forecast_date_date": "2022-05-16",
            "forecast_date_time": "00:00:00",
            "miss_val": "nan",
            "station_name": "PDLC2R",
            "lat": 38.4872222222,
            "lon": -105.373055556,
            "x": null,
            "y": null,
            "z": null,
            "units": "CFS",
            "creation_date": "2022-05-16",
            "creation_time": null
        },
</code>
```

Its providing only metadata, and my URL limited it to the first 10 headers.

Let me take a look at the Swagger,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-11-02T17:47:21Z

---

Not a big fan of the Swagger documentation. It may not be fully baked. I've asked David Moring to provide an example URL for a specific scenario that I can use as a starting point for understanding the service.

Thanks,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-11-02T17:47:39Z

---

Waiting for a reply,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-11-02T18:51:28Z

---

Was told in the WRES/WRDS meeting that they will work that request into examples dovetailed into a Jupyter Notebook they are hoping to release next Wednesday. I also informed them about the lack of a time zone in the response, so that we can't be sure in what time zone dates are presented.

Last point... They intend to move all services to something like the HEFS approach, with the next one likely being the Location Service, which will be moved to the Cloud.

Leaving on hold,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-11-16T21:01:53Z

---

Here is the Jupyter Notebook (enter the token "WRDSJupyter"):

https://nwcal-wrds.[host]/jupyter/lab/tree/hefs

It should help to investigate how to use the HEFS service. We would then need to write a brand new reader to prepare proper requests and parse the results.

Hank

**Comment by HankHerr-NOAA** [comment]
Latest from Gautam chatted to me today (edited to clean it up)...

HEFS is up and running in UAT. This is going to be our new landing page in the cloud (not pushed up to the cloud yet but the links work):

https://nwcal-wrds-dev01.[domain omitted]/.

heck out this notebook.  it shows how to retrieve an ensemble for a location:

https://colab.research.google.com/github/NOAA-OWP/data-service-notebooks/blob/master/HEFS/2_ensemble_plotting_demo.ipynb

I don't think we need to code against it, yet, but we should at least take a look before the next WRES/WRDS meeting (early November, since I canceled the meet tomorrow) so that we can discuss it if we have questions.

Thanks,

Hank

**Comment by james-d-brown** [comment]
Looking at this one.

**Comment by james-d-brown** [comment]
Starting with a small amount of refactoring to leverage a common utility for reading from a web source, i.e., wrapping the `WebClient` and handling http responses. This currently happens in various places using a `getInputStreamFromUri` method with very slight differences in handling of error responses across services that can be centralized. Also, more importantly, each reader uses a separate `WebClient`, and each `WebClient` wraps a separate `OkHttpClient`, but the advice from the authors of that client library is "OkHttp performs best when you create a single OkHttpClient instance and reuse it for all of your HTTP calls". So, doing that.

**Comment by james-d-brown** [comment]
> Starting with a small amount of refactoring to leverage a common utility for reading from a web source, i.e., wrapping the `WebClient` and handling http responses. This currently happens in various places using a `getInputStreamFromUri` method with very slight differences in handling of error responses across services, which can be centralized. Also, more importantly, each reader uses a separate `WebClient`, and each `WebClient` wraps a separate `OkHttpClient`, but the advice from the authors of that client library is "OkHttp performs best when you create a single OkHttpClient instance and reuse it for all of your HTTP calls". So, doing that.

Pushed this.

**Comment by HankHerr-NOAA** [comment]
I think David Moring might be the person to reach out to for that. You could even point him to this ticket, if you want.

I haven't taken a look and probably won't be able to today due to limited time,

Hank

**Comment by james-d-brown** [comment]
Yeah, I will definitely reach out, just wanted to make sure I wasn't missing something obvious before bothering them.

**Comment by james-d-brown** [comment]
Reached out and David is adding that capability.

On a separate topic, there are sufficient similarities between the data model adopted by the JSON formatted response from this service and the data model adopted by PI-XML that it's time to abstract out some of the deserialization from these two contexts to a common location and to re-use code and tests across these two contexts.

**Comment by james-d-brown** [comment]
> On a separate topic, there are sufficient similarities between the data model adopted by the JSON formatted response from this service and the data model adopted by PI-XML that it's time to abstract out some of the deserialization from these two contexts to a common location and to re-use code and tests across these two contexts.

Have the deserialization working for an example response. Need to abstract out the common aspects for pi-xml and this json format, as above, and then add a bunch of unit tests. Will then move on to the service aspects of formulating requests and teeing up responses for deserialization.

**Comment by james-d-brown** [comment]
Improving the test coverage for `ReaderUtilities`...

**Comment by james-d-brown** [comment]
Have a draft of the reader. To begin with, there will be no chunking by datetime range because the WRDS HEFS service does not properly support this yet. While the parameters are additive, so you could build a url with as many forecast datetime instances as you wanted, in principle, it could become impractically long and would probably fail validation for length (e.g., browsers, GET requests etc. are generally limited to 2048 characters). In short, it isn't currently viable, so you get "everything" w/r to time. It is chunked on location, though. Also, the service currently only serves 30 days per location so.... meh.

**Comment by james-d-brown** [comment]
Need to write a bunch of junit unit/integration tests for the reader class.

**Comment by james-d-brown** [comment]
More on that tomorrow.

**Comment by james-d-brown** [comment]
Have some unit tests.

Need to add some declaration validation. For example, the `variable` `name` must be present when the inferred or declared source interface is `WRDS_HEFS`.

**Comment by james-d-brown** [comment]
Have that validation done. Trying some real evaluations.

**Comment by james-d-brown** [comment]
Just discovered an inconsistent response format from the service. For example, consider this request URL:

https://api.water.noaa.gov/hefs/v1/ensembles/?location_id=PGRC2&parameter_id=QINE

It returns this:

```
[
  {
    "events": [
      {
        "date": "2025-03-18",
        "flag": "0",
        "time": "00:00:00",
        "value": 81.22373
      },
      {
        "date": "2025-03-
...
```

Which is not expected. When including a `limit` parameter, at least, there is a composition of `results`, not just a raw array of time-series.

**Comment by epag** [comment]
Also its worth mentioning that the way the results from this endpoint look may be changing in the future. I have a meeting with Gautam next week to discuss that

**Comment by HankHerr-NOAA** [comment]
My apologies for that. I was under the impression that the API was more "final" that it really was. It was perhaps too soon to work on this. Hopefully, the changes they make won't render useless any of the code you produced.

Hank

**Comment by james-d-brown** [comment]
Oh, not at all, it is no one's fault, it's just the nature of development against new services, I guess and, either way, it's all the same to me :)

**Comment by epag** [comment]
I think the current goal is to have everything API wise finalized by the end of the month, but I can let you know if there are any changes to that plan

**Comment by james-d-brown** [comment]
Thanks, Evan.

That's it for this week. Will come back to it next week. Main thing to do is to retrofit the `PublishedInterfaceXmlReader` to use the common abstractions developed for this one.

**Comment by james-d-brown** [comment]
Starting to retrofit the PI-XML reading stuff...

**Comment by james-d-brown** [comment]
Seems to be working fine. Here's an example over multiple features. Takes about 30 seconds, although this is obviously for very limited data, not even a month (more than a month requested, but not available).

```
vobserved:
  sources:
  - interface: usgs nwis
    uri: https://nwis.waterservices.usgs.gov/nwis/iv
  variable: '00060'
predicted:
  sources:
  - interface: wrds hefs
    uri: https://api.water.noaa.gov/hefs/v1/ensembles/
  variable: QINE
unit: ft3/s
valid_dates:
  minimum: '2025-04-01T06:00:00Z'
  maximum: '2025-05-13T11:00:00Z'
reference_dates:
  minimum: '2025-04-01T06:00:00Z'
  maximum: '2025-05-13T06:00:00Z'
lead_time_pools:
  period: 6
  unit: hours
lead_times:
  minimum: 0
  maximum: 120
  unit: hours
features:
- {observed: "01452000", predicted: "ALNP1"}
- {observed: "01502632", predicted: "BAIN6"}
- {observed: "01403060", predicted: "BDKN4"}
- {observed: "01548005", predicted: "BECP1"}
- {observed: "02026000", predicted: "BECV2"}
- {observed: "01453000", predicted: "BETP1"}
- {observed: "01402000", predicted: "BKWN4"}
- {observed: "01538700", predicted: "BMBP1"}
- {observed: "01503500", predicted: "BNGN6"}
- {observed: "02019500", predicted: "BNNV2"}
- {observed: "01381000", predicted: "BONN4"}
- {observed: "01380450", predicted: "BOTN4"}
- {observed: "01470500", predicted: "BREP1"}
- {observed: "01436690", predicted: "BRGN6"}
- {observed: "01646500", predicted: "BRKM2"}
- {observed: "01428500", predicted: "BRYN6"}
- {observed: "01446500", predicted: "BVDN4"}
- {observed: "02024000", predicted: "BVSV2"}
- {observed: "01379500", predicted: "CAMN4"}
- {observed: "02035000", predicted: "CARV2"}
- {observed: "01603000", predicted: "CBEM2"}
- {observed: "01427510", predicted: "CCRN6"}
- {observed: "01481000", predicted: "CDFP1"}
- {observed: "01548500", predicted: "CEDP1"}
metrics:
- bias fraction
- sample size
- box plot of errors by observed value
output_formats:
- csv2
- pairs
- png
```

**Comment by james-d-brown** [comment]
Finished retrofitting the pi-xml.

**Comment by james-d-brown** [comment]
Going to merge from main and push.

**Comment by james-d-brown** [comment]
Pushed.

**Comment by james-d-brown** [comment]
Assuming the build passes, this one should be good for UAT.

However, given that the API is being actively developed, working for now doesn't necessarily mean working on deployment of v6.32, so I will mark this as on hold. 

Will probably need a wiki entry too.

**Comment by HankHerr-NOAA** [comment]
Reopening, assigning to 6.33, and placing on-hold.

Hank

**Comment by HankHerr-NOAA** [comment]
The HEFS API is in UAT (I believe this is a cloud location):

 https://testing-api.water.noaa.gov/hefs/v1/docs/#/

It goes to production next week. We don't have any active requests to use the API, so this might not been high priority. The again, folks may not be asking about it, because they don't know about it.

Hank

**Comment by epag** [comment]
The HEFS API was deployed this past monday (08/28/2025) which updated the form of the responses. No breaking changes are currently planned so we could start to code against this

**Comment by epag** [comment]
Also the docs page is update to this now:

https://nwcal-wrds.[domain]/api/hefs/docs/

**Comment by epag** [comment]
One more note, the API is still going to be marked "EXPERIMENTAL" until sometime early next year I think, so there is some inherit volatility that may still be upcoming.

**Comment by epag** [comment]
The best way would be to email Gautam and myself currently

**Comment by james-d-brown** [comment]
I see. Personally, I would say that it's a bug that the forecast information can be served without promising that the basic information required to interpret it properly is present. I understand the desire to pass through information unchanged, but an api should make certain promises. As a consumer, I expect to see all of the basic information that is _necessary_ to qualify a forecast. Perhaps these forecasts need to be validated/filtered on ingest and a procedure established to feed invalid data back to the RFC for review.

The consequence of not making promises and having a much looser api is that downstream code needs to handle it and downstream code is always a great multiple of upstream code.

**Comment by epag** [comment]
Ill shoot him a message and I have a note to talk with him about this though

**Comment by epag** [comment]
I just sent him a link to this comment and I have a meeting with him tomorrow that I can discuss this with him

**Comment by james-d-brown** [comment]
Have this refactored and the existing unit tests are passing. Would benefit from a few more unit tests for sad paths, such as missing forecast information (which is currently seen in the wild). Also needs to be system tested, informally (I won't add a formal system test as this is a non-public api).

**Comment by james-d-brown** [comment]
Added a unit test for the sad path with a missing forecast datetime.

**Comment by james-d-brown** [comment]
Main issue with top-level integration testing wres w/r to the wrds production service is that it doesn't seem to contain any data. Looking at the testing instance instead.

**Comment by james-d-brown** [comment]
Everything seems to be working. Have tested a variety of shapes, including single features and multiple features. There is, however, very limited data available. A couple of example evaluations that succeed at the time of writing (but may not later).

One feature:

```
observed:
  sources:
  - interface: usgs nwis
    uri: https://nwis.waterservices.usgs.gov/nwis/iv
  variable: '00060'
predicted:
  sources:
  - interface: wrds hefs
    uri: https://testing-api.water.noaa.gov/hefs/v1/ensembles/
  variable: QINE
unit: ft3/s
valid_dates:
  minimum: '2025-09-01T00:00:00Z'
  maximum: '2025-10-20T00:00:00Z'
reference_dates:
  minimum: '2025-09-01T00:00:00Z'
  maximum: '2025-10-20T00:00:00Z'
lead_time_pools:
  period: 6
  unit: hours
lead_times:
  minimum: 0
  maximum: 120
  unit: hours
features:
- observed: '01570500'
  predicted: 'HARP1'
metrics:
- bias fraction
- sample size
- box plot of errors by observed value
output_formats:
- csv2
- pairs
- png
```

Multiple features:

```
observed:
  sources:
  - interface: usgs nwis
    uri: https://nwis.waterservices.usgs.gov/nwis/iv
  variable: '00060'
predicted:
  sources:
  - interface: wrds hefs
    uri: https://api.water.noaa.gov/hefs/v1/ensembles/
  variable: QINE
unit: ft3/s
valid_dates:
  minimum: '2025-10-01T06:00:00Z'
  maximum: '2025-10-31T11:00:00Z'
reference_dates:
  minimum: '2025-10-01T06:00:00Z'
  maximum: '2025-10-31T06:00:00Z'
lead_time_pools:
  period: 6
  unit: hours
lead_times:
  minimum: 0
  maximum: 120
  unit: hours
features:
- {observed: "01452000", predicted: "ALNP1"}
- {observed: "01502632", predicted: "BAIN6"}
- {observed: "01403060", predicted: "BDKN4"}
- {observed: "01548005", predicted: "BECP1"}
- {observed: "02026000", predicted: "BECV2"}
- {observed: "01453000", predicted: "BETP1"}
- {observed: "01402000", predicted: "BKWN4"}
- {observed: "01538700", predicted: "BMBP1"}
- {observed: "01503500", predicted: "BNGN6"}
- {observed: "02019500", predicted: "BNNV2"}
- {observed: "01381000", predicted: "BONN4"}
- {observed: "01380450", predicted: "BOTN4"}
- {observed: "01470500", predicted: "BREP1"}
- {observed: "01436690", predicted: "BRGN6"}
- {observed: "01646500", predicted: "BRKM2"}
- {observed: "01428500", predicted: "BRYN6"}
- {observed: "01446500", predicted: "BVDN4"}
- {observed: "02024000", predicted: "BVSV2"}
- {observed: "01379500", predicted: "CAMN4"}
- {observed: "02035000", predicted: "CARV2"}
- {observed: "01603000", predicted: "CBEM2"}
- {observed: "01427510", predicted: "CCRN6"}
- {observed: "01481000", predicted: "CDFP1"}
- {observed: "01548500", predicted: "CEDP1"}
metrics:
- bias fraction
- sample size
- box plot of errors by observed value
summary_statistics:
  dimensions:
    - features
  statistics:
    - mean
output_formats:
- csv2
- pairs
- png
```

There's around 10 or fewer forecasts being evaluated here, typically.

**Comment by james-d-brown** [comment]
Ready for UAT.

**Comment by epag** [comment]
Tested the single feature job and this passes UAT

**Comment by HankHerr-NOAA** [comment]
#688 created for better documentation of source formats and services.

Hank

**Comment by HankHerr-NOAA** [comment]
I asked Gautam if he's ready for folks to use the HEFS API via WRES. His response:

> of course, as long as they understand it's experimental and won't be considered operational till end of fy26

I'll add the word experimental in the News item.

Hank

**Comment by james-d-brown** [comment]
You can still visualize the errors in a time-series plot, for example. The capability is there, we should advertise it, but note the constraints.

**Comment by HankHerr-NOAA** [comment]
Ah, true. That time series plotting capability for short time periods has definitely proven useful for me (investigating an HEFS issue).

Hank
